feat(markdown): add support for `shiki` option `langAlias` (#12039)
withastro/astro: .changeset/dirty-socks-sip.md

---
'@astrojs/markdown-remark': minor
'astro': minor
---

> Adds a [`markdown.shikiConfig.langAlias` option](https://shiki.style/guide/load-lang#custom-language-aliases) that allows aliasing a non-supported code language to a known language.

Reviewer: ```suggestion
Adds a `markdown.shikiConfig.langAlias` option that allows [aliasing a non-supported code language to a known language](https://shiki.style/guide/load-lang#custom-language-aliases). This is useful when the language of your code samples is not [a built-in Shiki language](https://shiki.style/languages), but you want your Markdown source to display an accurate language while also displaying syntax highlighting.
```
Maybe something like this?

> Adds a configuration called https://shiki.style/guide/load-lang#custom-language-aliases, that allows a non-supported code language to a known language.
>  
> The below example will tell shiki to highlight the code blocks `cjs` using the `javascript` syntax highlighting.  

Reviewer: ```suggestion
Adds a new Shiki configuration option `langAlias` to use [Shiki's `langAlias` option](https://shiki.style/guide/load-lang#custom-language-aliases). This allows you to define custom aliases for [Shiki's bundled language IDs](https://shiki.style/languages#bundled-languages) (such as the built-in `js` for `javascript`) for Markdown code blocks.

This can be used to map a non-supported code language to a known language, allowing you to indicate the exact language in your source code while using appropriate highlighting.
 
The example below configures any code blocks using the language alias `cjs` to use `javascript` syntax highlighting.  
```
Maybe something like this? I took a look at the original feature, and their example seems a little more general purpose than only using a language that doesn't have a supported code. It seems you could create an alias for any reason, using anything, so I think describing this as one particular use case vs. "what this thing is for" frames it as being able to be used in other ways, too.

Author: @sarah11918 I updated the changeset based on https://github.com/withastro/astro/pull/12039#discussion_r1768629167, can you provide a new suggestion?

> For example, the below configuration tells shiki to highlight `cjs` code blocks using the `javascript` syntax highlighter:

Reviewer: ```suggestion
The following example configures Shiki to highlight `cjs` code blocks using the `javascript` syntax highlighter:
```

```js
import { defineConfig } from 'astro/config';

export default defineConfig({
  markdown: {
    shikiConfig: {
      langAlias: {
        cjs: 'javascript',
      },
    },
  },
});
```

> 

Reviewer: ```suggestion

Then in your Markdown, you can use the alias as the language for a code block for syntax highlighting:

```

---
'@astrojs/markdown-remark': minor
'astro': minor
---

> Adds a `markdown.shikiConfig.langAlias` option that allows [aliasing a non-supported code language to a known language](https://shiki.style/guide/load-lang#custom-language-aliases). This is useful when the language of your code samples is not [a built-in Shiki language](https://shiki.style/languages), but you want your Markdown source to display an accurate language while also displaying syntax highlighting.

Reviewer: ```suggestion
Adds a `markdown.shikiConfig.langAlias` option that allows [aliasing a non-supported code language to a known language](https://shiki.style/guide/load-lang#custom-language-aliases). This is useful when the language of your code samples is not [a built-in Shiki language](https://shiki.style/languages), but you want your Markdown source to contain an accurate language while also displaying syntax highlighting.
```
Sorry, just noticed "display" twice in the same sentence! This is better 😄

---
'@astrojs/markdown-remark': minor
'astro': minor
---

Adds a configuration called https://shiki.style/guide/load-lang#custom-language-aliases, that allows a non-supported code language to a known language.

> This option requires `langs` to be defined with the correct values. The option will tell shiki which language to load when mapping the alias.

Reviewer 2: Digging deeper into this now, it seems like a Shiki bug that requires you to specify the `langs` beforehand for aliases. It might be good to followup upstream later.

Author: Well, the docs show that `javascript` is the array

https://shiki.style/guide/load-lang#custom-language-aliases

Reviewer 2: I don't think you have to. You can specify the langs upfront so that you can syntax highlight the code synchronously, but Astro doesn't do that and will always load the langs dynamically, so in practice the `langs` field is only for additional custom languages.

Reviewer: Just noting that I came to peek in on this PR and I'm not sure whether this behaviour has been confirmed yet, so I'll wait to review!

Reviewer 2: Is good for docs review now 👍
